### PR TITLE
refactor: VLA cleanup part 1.

### PR DESCRIPTION
**Pull request by crypto-universe**
Replacing VLAs either by malloc/calloc or by static stack allocation.
See https://github.com/TokTok/c-toxcore/issues/959

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1263)

### CODE REVIEW

**Comment by crypto-universe** [comment]
Hm, on my local run all test are passed except `lan_discovery`, which hangs even on clean master...

File: auto_tests/messenger_test.c [hunk 1]
```diff
@@ -232,7 +232,7 @@ START_TEST(test_dht_state_saveloadsave)
      * d) the second save() is of equal content */
     const size_t extra = 64;
     const size_t size = dht_size(m->dht);
-    VLA(uint8_t, buffer, size + 2 * extra);
+    uint8_t* const buffer = malloc((size + 2 * extra) * sizeof(uint8_t));
```
Comment (iphydf) [inline auto_tests/messenger_test.c]: Either `uint8_t *const buffer = (uint8_t *)calloc(size ..., sizeof(uint8_t));` or `uint8_t *const buffer = (uint8_t *)malloc(size + 2 * extra);` without the sizeof. If you use sizeof, use calloc.

Same in other places.

File: auto_tests/messenger_test.c [hunk 2]
```diff
@@ -258,10 +258,12 @@ START_TEST(test_dht_state_saveloadsave)
     ck_assert_msg(size == size2, "Messenger \"grew\" in size from a store/load cycle: %u -> %u", (unsigned)size,
                   (unsigned)size2);
 
-    VLA(uint8_t, buffer2, size2);
+    uint8_t* const buffer2 = malloc(size2 * sizeof(uint8_t));
```
Comment (iphydf) [inline auto_tests/messenger_test.c]: Like here.

File: auto_tests/tox_one_test.c [hunk 3]
```diff
@@ -77,8 +76,8 @@ static void test_one(void)
                   "Can't set status message of TOX_MAX_STATUS_MESSAGE_LENGTH");
 
     tox_self_get_address(tox1, address);
-    size_t save_size = tox_get_savedata_size(tox1);
-    VLA(uint8_t, data, save_size);
+    const size_t save_size = tox_get_savedata_size(tox1);
+    uint8_t* data = malloc(save_size);
```
Comment (iphydf) [inline auto_tests/tox_one_test.c]: `uint8_t *const data`

File: toxav/rtp.c [hunk 4]
```diff
@@ -788,22 +788,22 @@ int rtp_send_data(RTPSession *session, const uint8_t *data, uint32_t length,
         header.flags |= RTP_KEY_FRAME;
     }
 
-    VLA(uint8_t, rdata, length + RTP_HEADER_SIZE + 1);
-    memset(rdata, 0, SIZEOF_VLA(rdata));
+    const size_t packet_size = length + RTP_HEADER_SIZE + 1;
+    uint8_t* rdata = calloc(packet_size, sizeof(uint8_t));
```
Comment (iphydf) [inline toxav/rtp.c]: `uint8_t *const rdata`

File: toxav/toxav.c [hunk 5]
```diff
@@ -769,24 +769,26 @@ bool toxav_audio_send_frame(ToxAV *av, uint32_t friend_number, const int16_t *pc
             goto RETURN;
         }
 
-        VLA(uint8_t, dest, sample_count + sizeof(sampling_rate)); /* This is more than enough always */
+        uint8_t* dest = malloc(sample_count + sizeof(sampling_rate)); /* This is more than enough always */
```
Comment (iphydf) [inline toxav/toxav.c]: Move the `*`, and add `const`.

File: toxcore/TCP_connection.c [hunk 6]
```diff
@@ -1467,7 +1467,7 @@ static void kill_nonused_tcp(TCP_Connections *tcp_c)
     unsigned int i;
     unsigned int num_online = 0;
     unsigned int num_kill = 0;
-    VLA(unsigned int, to_kill, tcp_c->tcp_connections_length);
+    unsigned int* const to_kill = malloc(tcp_c->tcp_connections_length * sizeof(unsigned int));
```
Comment (iphydf) [inline toxcore/TCP_connection.c]: `calloc` and move `*`.

File: toxcore/crypto_core.c [hunk 7]
```diff
@@ -142,20 +141,23 @@ int32_t encrypt_data_symmetric(const uint8_t *secret_key, const uint8_t *nonce,
         return -1;
     }
 
-    VLA(uint8_t, temp_plain, length + crypto_box_ZEROBYTES);
-    VLA(uint8_t, temp_encrypted, length + crypto_box_MACBYTES + crypto_box_BOXZEROBYTES);
+    uint8_t* temp_plain = calloc(length + crypto_box_ZEROBYTES, sizeof(uint8_t));
+    uint8_t* temp_encrypted = malloc(length + crypto_box_MACBYTES + crypto_box_BOXZEROBYTES);
```
Comment (iphydf) [inline toxcore/crypto_core.c]: Please be consistent. Choose one, and do that everywhere (probably calloc?).

File: toxcore/net_crypto.c [hunk 8]
```diff
@@ -1073,21 +1073,25 @@ static int send_data_packet(Net_Crypto *c, int crypt_connection_id, const uint8_
         return -1;
     }
 
+    const size_t packet_size = 1 + sizeof(uint16_t) + length + CRYPTO_MAC_SIZE;
+    uint8_t* const packet = calloc(packet_size, sizeof(uint8_t));
```
Comment (iphydf) [inline toxcore/net_crypto.c]: `uint8_t *const p..`

File: toxcore/net_crypto.c [hunk 9]
```diff
@@ -1105,13 +1109,16 @@ static int send_data_packet_helper(Net_Crypto *c, int crypt_connection_id, uint3
     num = net_htonl(num);
     buffer_start = net_htonl(buffer_start);
     uint16_t padding_length = (MAX_CRYPTO_DATA_SIZE - length) % CRYPTO_MAX_PADDING;
-    VLA(uint8_t, packet, sizeof(uint32_t) + sizeof(uint32_t) + padding_length + length);
+    const size_t packet_size = sizeof(uint32_t) + sizeof(uint32_t) + padding_length + length;
+    uint8_t* const packet = calloc(packet_size, sizeof(uint8_t));
```
Comment (iphydf) [inline toxcore/net_crypto.c]: `uint8_t *const`

File: toxcore/util.c [hunk 10]
```diff
@@ -55,14 +53,16 @@ uint32_t id_copy(uint8_t *dest, const uint8_t *src)
...
-    uint32_t i;
-    VLA(uint8_t, buff, numbytes);
-
-    for (i = 0; i < numbytes; ++i) {
-        buff[i] = num[numbytes - i - 1];
+    const size_t num_of_iterations = numbytes / 2;
+    size_t i;
+    uint8_t temp;
+
+    for (i = 0; i < num_of_iterations; ++i) {
```
Comment (iphydf) [inline toxcore/util.c]: `for (size_t i = 0; ...`

**Comment by iphydf** [comment]
Can you fix the build failures in this PR?

**Comment by crypto-universe** [comment]
> Can you fix the build failures in this PR?

I have a feeling that CI fails totally randomly. I'll do my best to fix builds, but there are no guarantees.

**Comment by iphydf** [comment]
https://travis-ci.org/TokTok/c-toxcore/jobs/455236334#L694 needs to be fixed, that's not random.

File: toxcore/TCP_connection.c (hunk 6 above)
Comment (crypto-universe) [reply toxcore/TCP_connection.c]: I use calloc where VLA was followed by zeroing memory via `memset`. In other cases malloc (for performance reasons.

File: toxcore/crypto_core.c (hunk 7 above)
Comment (crypto-universe) [reply toxcore/crypto_core.c]: Not relevant any more. This place was fixed by zoff99.

**Comment by iphydf** [comment]
@crypto-universe can you let me know when you think you've fixed the CI errors to the point where it's no longer fixable in this PR, i.e. it becomes random flakiness?

**Comment by crypto-universe** [comment]
@iphydf 
I thought that if size of array is known at compile time, it won't be VLA. Sometimes I really lack constexpr and other features from modern C++ :(
I see two options, both are bad:
1. Mallocs everywhere. Slow solution, leads to heap memory fragmentation.
2. Static arrays, global for translation unit (across C file). Fast, but violates good practices and not thread safe.
`enum` in my opinion will be less readable.
Due to `calloc`... I think both methods are well-known and readable. `calloc` is crucial when allocating memory for structures, because it will zero padding bytes inside structure and no info will leak via those chunks of memory.
Normally I build (on Linux) and test my commits before pushing. So it is veeery strange, that CI fails.

**Comment by iphydf** [comment]
In old C++98 that would also be a constant expression. It's just C that doesn't do this (sadly). Don't do static arrays. A third option is to implement the bump-allocator I made an API for the other day. Basically, malloc and free would just bump a pointer in a large pre-allocated array. We'd use a linked list of large memory blocks to allocate from. It would be a stack-like memory allocator next to the runtime stack. I intend to use that anyway at some point (region/arena-based allocation). It avoids malloc for lexically bound allocations. In this PR, I would either use malloc/calloc or make an API for bump allocator and use it, but implement it with plain malloc, deferring the allocator implementation to a separate PR.

**Comment by crypto-universe** [comment]
@iphydf what do you think about such solution?
```
int funcB()
{
  static int ints[1024]; /* this is "static", i.e. the memory "survive" even
                             when the function exits, but the symbol's scope is local */ 
...........
}
```

From https://rosettacode.org/wiki/Memory_Allocation#C

**Comment by iphydf** [comment]
That's almost exactly the same as `static int funcB_ints[1024]; int funcB() { ... }` so doesn't solve any of the problems global variables have.

**Comment by crypto-universe** [comment]
It limits visibility scope to the method where it is defined. This array will be super private.

**Comment by iphydf** [comment]
To be precise, it's almost exactly the same as `static int ints.1796[1024]`, declaring a global symbol with the name `ints.1796` (or some other number). It has the same thread-safety issues as namespace-scope globals. It is a global variable, even if its scope is limited.

**Comment by crypto-universe** [comment]
On thread safety totally agree.
On the other hand, custom "allocator from a huge preallocated chunk of memory" will have same problem.
Solve it with mutexes... Looks like an overkill. I'd prefer `malloc` way to custom allocator with mutex.

By the way, what is a current status of toxcore: does it use multithreading? As for me, possible multithreading issues should be addressed when they appear...

**Comment by iphydf** [comment]
@crypto-universe toxcore has some thread safety guarantees that are described in the documentation (tox.h). In particular, we guarantee that two instances of toxcore can run in parallel in different threads. Accesses to a single tox instance must be sequential.

File: toxcore/friend_requests.c [hunk 11]
```diff
@@ -158,10 +158,10 @@ static int friendreq_handlepacket(void *object, const uint8_t *source_pubkey, co
 
     addto_receivedlist(fr, source_pubkey);
 
-    const uint32_t message_len = length - sizeof(fr->nospam);
-    VLA(uint8_t, message, message_len + 1);
+    const size_t message_len = (size_t)length - sizeof(fr->nospam);
+    uint8_t message[ONION_CLIENT_MAX_DATA_SIZE + 1];
     memcpy(message, packet + sizeof(fr->nospam), message_len);
-    message[SIZEOF_VLA(message) - 1] = 0; /* Be sure the message is null terminated. */
+    message[message_len] = 0; /* Be sure the message is null terminated. */
```
Comment (iphydf) [inline toxcore/friend_requests.c]: This is a VLA. Even `constexpr` wouldn't save you here, because `length` is not a constexpr.

**Comment by iphydf** [comment]
@crypto-universe this PR needs rebase.

**Comment by crypto-universe** [comment]
Hi, @iphydf ! PR is one year old. I think it is easier to close it and create a new one, from scratch on fresh codebase. What do you think?

**Comment by iphydf** [comment]
Yes, I agree :). Maybe we should talk about how to best handle packet allocations in general. The advantage of VLA/alloca is that

1) only the necessary amount of memory is allocated (as opposed to allocating the maximum packet size)
2) the memory is local to the other parts of the local program memory (cache locality)
3) allocating and deallocating it is *very* fast (compared to malloc/free).

We could achieve most of those goals in a different way, but using malloc/free directly isn't that way. We should check whether it's worth doing something extra to keep these properties, or if malloc/free is actually sufficient for our use case (it probably is).

**Pull request closed without merging**